File: passctl/cmds/add.py
from cmds import Command
from log import error, inp, success, info 
from random import choice
from getpass import getpass
import readline
import string

class Add(Command):
    def __init__(self) -> None:
        super().__init__(
            "add", 
            "Add an entry to the vault",
        )

    def run(self) -> None:
        if len(self.args)<1:
            error("Entry name not specified")

        entry = self.args[0]
        vault = self.stg.data

        if entry in vault.keys():
            error("Entry already exists")
       
        try:
            info("Leave the password empty for auto generation")
            usr = inp("Username")
            pwd = inp("Password", pwd=True)
        except KeyboardInterrupt:
            error("Aborted", start="\n")
            return 

        if pwd == "":
            chars = string.ascii_letters+string.digits+"!?=%.:$+-"
            for _ in range(20):
                pwd += choice(chars)
            success(f"Auto generated password: {pwd}")
        
        vault[entry] = {
            "user": usr, 
            "pass":pwd
        }
        success(f"Added {entry}")

        self.stg.data = vault
        if not self.stg.write():
            error("Cannot save the vault")

File: cmds/add.go
package cmds

import (
	"os"

	"github.com/passctl/client/config"
	"github.com/passctl/client/lib"
)

func Add() bool {
  if len(os.Args) != 3 {
    lib.Error("Specify a name for the new entry")
    return false
  }

	if config.Cfg.Token == "" {
    lib.Error("You should first run the %ssetup%s command!",
      lib.COLOR_BOLD, lib.COLOR_RESET)
    return false
  } 

  var entry config.Entry
  entry.Name = os.Args[2]
  if len(entry.Name) < 3 {
    lib.Error("Entry name should be at least %s3%s characters long",
      lib.COLOR_BOLD, lib.COLOR_RESET)
    return false
  }

  master := GetMaster()

USER:
  err := config.LoadEntries(master)
  if err != nil {
    lib.ErrorMsg("Failed to load entries", err)
    return false
  }

  if config.GetEntry(entry.Name)!=nil {
    lib.Error("An entry with the same name exists") 
    return false
  }

  entry.User = lib.Input("Enter a username")
  if entry.User == "" {
    lib.Error("Username cannot be empty")
    goto USER
  }

  lib.Info("Leave the password field empty for auto generation")
  entry.Pwd = lib.Passwd("Enter a password")

  if entry.Pwd == "" {
    entry.Pwd = GenPwd(12)
    lib.Success("Generated password: %s%s", 
      lib.COLOR_BOLD, entry.Pwd)
  }

  config.Entries = append(config.Entries, entry)
  err = config.SaveEntries(master)
  if err != nil {
    lib.ErrorMsg("Failed to save entries", err)
    return false
  }

  err = config.Save()
  if err != nil {
    lib.ErrorMsg("Failed to save configuration", err)
    return false
  }

  lib.Success("Configuration has been saved")
  return true
}
